src/ip/Address.cc

```diff
@@ -695,9 +696,11 @@ Ip::Address::InitAddr(struct addrinfo *&ai)
 void
 Ip::Address::FreeAddr(struct addrinfo *&ai)
 {
-    if (ai == nullptr) return;
+    if (!ai)
+        return;
 
-    if (ai->ai_addr) delete ai->ai_addr;
+    if (ai->ai_addr)
+        delete static_cast<struct sockaddr*>(struct sockaddr_in6*>(ai->ai_addr);
```

Comment:
I share Francesco's concern. Unfortunately, even more (than suggested) changes are needed to do this right because the current official code _sometimes_ does not set ai_addrlen before or immediately after allocating ai_addr. We need to refactor _all_ constructing code, so that we can rely on ai_addrlen during destruction, and I strongly recommend encapsulating this ai_addr memory allocation/destruction logic into a pair of dedicated functions.

I found three places where ai_addr memory is allocated and two places where it is freed. Please double check and keep all of them in sync.